File: kmeans1.scala
//Sample code of scala (for spark-shell)

val rdata = spark.read.csv("/devicestatus_etl/*")
val rdatac1 = rdata.select("_c3", "_c4").withColumnRenamed("_c3","lat").withColumn("lat", col("lat").cast("double")).withColumnRenamed("_c4","lon").withColumn("lon", col("lon").cast("double"))
val rdatac = rdatac1.where(($"lat" > 0))
//val rdatac = rdatac1.where((rdatac1.lat > 0))
val featuresSelected = ("lat","lon")

import org.apache.spark.ml.feature.VectorAssembler
import org.apache.spark.ml.linalg.Vectors
val va = new VectorAssembler().setInputCols(Array("lat","lon")).setOutputCol("features")
val rdataAssembled = va.transform(rdatac)

import org.apache.spark.ml.feature.StandardScaler
val ss = new StandardScaler().setInputCol("features").setOutputCol("scaledFeatures").
setWithStd(true).setWithMean(false)
val rdataStandardized=ss.fit(rdataAssembled).transform(rdataAssembled)

import org.apache.spark.ml.evaluation.ClusteringEvaluator
import org.apache.spark.ml.clustering.KMeans
val kmeans = new KMeans().setK(5).setSeed(1L)
val kmeansModel = kmeans.fit(rdataStandardized)
kmeansModel.computeCost(rdataStandardized)

println("Cluster Centers: ")
kmeansModel.clusterCenters.foreach(println)

File: kmeans1.py
# Sample code of pyspark
from __future__ import print_function

import numpy as np
import pandas as pd
#import matplotlib.pyplot as plt

#import seaborn as sns

#import folium 
from pyspark.ml.clustering import KMeans
from pyspark.sql.functions import col
from pyspark.ml.clustering import KMeans


#print (kmeans.explainParams())

rdata = spark.read.csv("/test/devicestatus_etl/*")
#rdata.printSchema()

rdatac1 = rdata.select("_c3", "_c4")\
        .withColumnRenamed("_c3","lat").withColumn("lat", col("lat").cast("double"))\
        .withColumnRenamed("_c4","lon").withColumn("lon", col("lon").cast("double"))

# FIXME: 0のデータをスキップする
rdatac = rdatac1.where((rdatac1.lat > 0))

#rdatac.printSchema()
#rdatac.show(3)
#rdata.take(3)

featuresSelected = ["lat","lon"]

from pyspark.ml.feature import VectorAssembler
va = VectorAssembler(inputCols=featuresSelected,outputCol="features")

rdataAssembled = va.transform(rdatac)


from pyspark.ml.feature import StandardScaler
ss = StandardScaler(inputCol="features",outputCol="featuresScaled", withMean=False,withStd=True)
rdataStandardized=ss.fit(rdataAssembled).transform(rdataAssembled)

#rdataStandardized.describe().show()

############################################################################
## 上記までがデータの準備。SparkのProgramming Guideに記載されている内容は
## libsvm形式のデータを読み込んでいますが、この例では devicestatus_etl
## 以下のデータを読み込んで変換しています

kmeans = KMeans(featuresCol="features",predictionCol="cluster", k=5)

#type(kmeans)
#print(kmeans.explainParams())

kmeansModel = kmeans.fit(rdataStandardized)
kmeansModel.computeCost(rdataStandardized)

centers = kmeansModel.clusterCenters()

print ("cluster centers")

for center in centers:
    print (center) 
